src/sentry.ts

```diff
@@ -23,20 +28,40 @@ const PINO_TO_SENTRY: { [key: number]: Severity } = {
     60: Severity.Critical,
 };
 
-export const createSentryTransformStream = (options: CosmasOptions): any => {
-    return class SentryTransformStream extends Transform {
-        // tslint:disable-next-line:function-name
-        public _transform(chunk: any, _encoding: string, callback: TransformCallback) {
-            const obj = JSON.parse(chunk);
-            if (obj.level >= (options.sentryLevel || levels.warn)) {
-                withScope((scope) => {
-                    scope.setLevel(PINO_TO_SENTRY[obj.level]);
-                    scope.setExtras(obj);
-                    reportToSentry(obj);
-                });
-            }
-            this.push(chunk);
-            callback();
+export const extendSentry = (logger: Cosmas, options: { sentry: string | true; sentryLevel: number }) => {
+    const sentry = require('@sentry/node');
+    if (typeof options.sentry === 'string') {
+        sentry.init({ dsn: options.sentry });
+    }
+
+    const originalWrite = (logger as any)[pino.symbols.streamSym].write;
+    // unfortunately, this is the only place in pino, we can hook onto, where we can be sure all
+    // the hooks, formatters and serializers are already applied
+    (logger as any)[pino.symbols.streamSym].write = function (s: string) {
+        originalWrite.call(this, s);
+        const obj = JSON.parse(s);
+        if (obj.level < (options.sentryLevel || levels.warn)) return;
+        const sentryCallback: SentryCallback | undefined = clsNamespace.get('sentryCallback');
+        withScope((scope) => {
+            scope.setLevel(PINO_TO_SENTRY[obj.level]);
+            scope.setExtras(obj);
+            if (sentryCallback) sentryCallback(scope);
+            reportToSentry(obj);
+        });
+    };
+
+    logger.realHooks.logMethod = function (inputArgs, method) {
+        // 2nd argument is Sentry options
+        const sentryCallback: SentryCallback | undefined = inputArgs[1];
+        // TODO: automatic type for 2nd log method arg
+
+        if (!sentryCallback) {
+            return method.apply(this, inputArgs);
         }
+
+        clsNamespace.runAndReturn(() => {
+            clsNamespace.set('sentryCallback', sentryCallback);
+            return method.apply(this, inputArgs);
+        });
```

Comment:
Ok, so to get the scope function down to hooks, we use CLS. How does this compare to embedding the object in a reserved key and setting scope from those keys? Aren't you concerned about performance of using CLS?

Reply:
In the end, I used CLS because using reserved key, we would interfere with pino's internal logic. Pino has separate processings for
a) you pass a string
b) you pass an object

If we would use a reserved keys, we would convert all the A cases to B cases, thus interfering with how pino process the arguments and creates a final "object to be logged" from them.

Reply:
OK